### PR TITLE
Shortest Vector Feature: 2D displacement to nearest foil surface

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**GLOBAL GEOMETRY DESCRIPTOR from FVF paper (ICML 2024).** The current DSDF gives SCALAR distances to each foil separately (dsdf1, dsdf2). The Shortest Vector (SV) representation gives the 2D VECTOR [dx, dy] to the nearest surface point ACROSS ALL FOILS — encoding both which foil is closest AND in what direction.

For tandem configs, nodes in the inter-foil slot have similar DSDF to both foils. The model must disambiguate "am I closer to the pressure side of the aft foil or the suction side of the fore foil?" from two separate DSDF scalars. The SV makes this geometrically unambiguous with a single 2D vector.

**Literature:** Lam et al. "Finite Volume Features, Global Geometry Representations, and Residual Training" (ICML 2024, arXiv:2402.02367) — SV representation improved MSE by 18-32% on AirfRANS and ShapeNet beyond DSDF alone.

## Instructions

Add `--shortest_vector_feature` flag. 2 new input channels: (sv_x, sv_y) — displacement vector from each node to the nearest surface point across all foils.

### Implementation

```python
def compute_shortest_vector(pos, surface_pos_foil1, surface_pos_foil2, is_tandem):
    """Shortest vector from each node to nearest foil surface point.
    pos:              [B, N, 2] all node positions
    surface_pos_foil1: [B, S1, 2] foil 1 surface positions
    surface_pos_foil2: [B, S2, 2] foil 2 surface positions (or zeros)
    is_tandem:        [B] boolean
    Returns:          [B, N, 2] shortest vector to nearest surface
    """
    # Distance to foil 1 surface: use torch.cdist for batch efficiency
    dist1 = torch.cdist(pos, surface_pos_foil1)  # [B, N, S1]
    min_dist1, min_idx1 = dist1.min(dim=-1)       # [B, N]
    
    # Gather the actual displacement vector to nearest foil1 surface point
    idx1_exp = min_idx1.unsqueeze(-1).expand(-1, -1, 2)  # [B, N, 2]
    nearest1 = surface_pos_foil1.gather(1, idx1_exp)      # [B, N, 2]
    sv1 = nearest1 - pos                                   # [B, N, 2] displacement
    
    if is_tandem.any():
        dist2 = torch.cdist(pos, surface_pos_foil2)       # [B, N, S2]
        min_dist2, min_idx2 = dist2.min(dim=-1)
        idx2_exp = min_idx2.unsqueeze(-1).expand(-1, -1, 2)
        nearest2 = surface_pos_foil2.gather(1, idx2_exp)
        sv2 = nearest2 - pos
        
        # Choose whichever foil is closer (per node)
        use_foil1 = (min_dist1 <= min_dist2).unsqueeze(-1)  # [B, N, 1]
        sv = torch.where(use_foil1, sv1, sv2)
    else:
        sv = sv1
    
    # Normalize by chord length for scale invariance
    chord = 1.0  # normalized
    sv_norm = sv / chord
    
    return sv_norm  # [B, N, 2]
```

### Key implementation details

1. **Surface node identification:** Reuse the surface mask from existing DSDF/SRF computation. Extract surface positions for each foil separately. Check `prepare_multi.py` for how foil1 vs foil2 surface nodes are separated.

2. **Compute ONCE per forward pass (not per epoch):** `torch.cdist` with N~10K nodes × S~300 surface nodes = 3M distances — fast on GPU. Cache if possible, but given augmentation changes geometry each epoch, compute at forward time.

3. **Concatenate to input:** `x = torch.cat([x, sv_norm], dim=-1)  # +2 channels`

4. **For single-foil samples:** `sv = sv1` (only one foil surface exists).

### Flags
```python
parser.add_argument('--shortest_vector_feature', action='store_true',
                    help='Add 2D shortest vector to nearest foil surface (FVF-paper global geometry)')
```

### Training commands
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --seed 42 \
  --wandb_name "frieren/shortest-vector-s42" \
  --wandb_group "shortest-vector-feature" \
  --shortest_vector_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73
```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Reference:** Lam et al. (ICML 2024, arXiv:2402.02367)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Seed 42
| Metric | Value |
|--------|-------|
| val_in_dist (p_in) | mae_surf_p=12.4, loss=0.2607 |
| val_tandem_transfer (p_tan) | mae_surf_p=28.6, loss=0.8868 |
| val_ood_cond (p_oodc) | mae_surf_p=8.0, loss=0.2047 |
| val_ood_re (p_re) | mae_surf_p=6.6, loss=0.2006 |
| Peak VRAM | 46.4 GB |
| Epoch time | ~77s |
| Epochs completed | 139 (180 min timeout) |
| W&B run | `42rldvb9` |

**Command:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/shortest-vector-s42" --wandb_group "shortest-vector-feature" \
  --shortest_vector_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

### Seed 73
| Metric | Value |
|--------|-------|
| val_in_dist (p_in) | mae_surf_p=11.8, loss=0.2478 |
| val_tandem_transfer (p_tan) | mae_surf_p=28.5, loss=0.8866 |
| val_ood_cond (p_oodc) | mae_surf_p=8.1, loss=0.2078 |
| val_ood_re (p_re) | mae_surf_p=6.8, loss=0.2020 |
| Peak VRAM | 45.2 GB |
| Epoch time | ~77s |
| Epochs completed | 139 (180 min timeout) |
| W&B run | `j1xftj9h` |

### 2-Seed Average vs Baseline

| Metric | Baseline (2-seed avg) | SV Feature (2-seed avg) | Delta |
|--------|----------------------|------------------------|-------|
| **p_in** | **11.742** | **12.10** | **+0.36 (3.1% worse)** |
| p_oodc | 7.643 | 8.05 | +0.41 (5.4% worse) |
| **p_tan** | **27.874** | **28.55** | **+0.68 (2.4% worse)** |
| p_re | 6.419 | 6.70 | +0.28 (4.4% worse) |

### What happened

**NEGATIVE RESULT.** The Shortest Vector feature did not improve any metric. All surface pressure MAE values are 2-5% worse than baseline across both seeds consistently.

**Why it likely failed:**

1. **Redundancy with existing features.** The model already has DSDF (8 channels giving signed distance fields to both foils), TE coordinate frame features (6 channels), and wake deficit features (2 channels). The SV's 2D displacement vector to the nearest surface point likely doesn't add meaningfully new geometric information beyond what DSDF already encodes — especially since DSDF has per-foil directional gradients that implicitly capture surface proximity and direction.

2. **Implementation limitations.** The `compute_shortest_vector` function uses a per-batch Python loop with `@torch.compiler.disable` to handle variable surface node counts. This prevents the function from being compiled and optimized. The subsampling to 512 surface points also introduces noise. While this didn't cause instability, it may have slightly degraded gradient quality.

3. **Different domain from the original paper.** Lam et al. (ICML 2024) tested SV on AirfRANS (single airfoil) and ShapeNet (diverse 3D shapes) where the model had minimal geometric context. Our model already has rich geometry features (DSDF, TE coords, wake deficit, SAF, curvature, distance-to-surface). The marginal value of SV is much lower when the model already has strong geometric priors.

4. **No normalization issue this time.** Unlike the stagnation pressure feature (PR #2291), the SV displacements are naturally in the same scale as the node positions, so there was no normalization instability. The feature simply wasn't useful.

### Suggested follow-ups

- **Signed normal distance:** Instead of the full 2D SV vector, try a single signed distance-to-nearest-surface feature that encodes which "side" of the nearest foil a node is on (positive for pressure side, negative for suction side). This disambiguates the inter-foil gap more directly.
- **Foil-specific SV:** Instead of nearest-across-all-foils, compute separate SVs to foil1 and foil2 surface, giving the model explicit multi-foil proximity. Though this may overlap even more with DSDF.
- **Focus on features that add genuinely new physics** rather than more geometry. The model's geometry encoding appears sufficient; the bottleneck may be in physics representation (e.g., local flow gradients, pressure recovery indicators).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2304 because ADVISOR: Closing — all metrics 2-5% worse. 10th consecutive failure.

**Your analysis is exactly right:** The model's geometry encoding (DSDF 8-ch, TE coords 6-ch, wake deficit 2-ch, SAF, curvature, distance-to-surface) is SATURATED. The Shortest Vector adds no new geometric information beyond what DSDF already encodes.

**Updated conclusion for the research programme:** We have now exhausted ALL reasonable input feature additions — every geometric/physics feature that's derivable from or complementary to existing inputs has failed. The bottleneck is NOT in the model's input representation.

The next set of experiments will focus entirely on training dynamics, regularization, and architectural efficiency — not new features. New assignment coming..
